docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+	Templates []Template `json:"templates"`
+}
+
+// AllowedTarget represents an allowed target.
+type AllowedTarget struct {
+	// Group is the group of the target resource.
+	Group gatewayv1.Group `json:"group"`
+
+	// Kind is kind of the target resource.
+	Kind gatewayv1.Kind `json:"kind"`
+}
+
+// Template is an NGINX configuration template.
+type Template struct {
+	// Context is the NGINX context to insert the template into.
+	Context NginxContext `json:"context"`
+
+	// Value is the template. It must be a valid Go template -- https://pkg.go.dev/text/template.
+	Value string `json:"value"`
+}
+
+// SnippetsTemplateStatus defines the state of the SnippetsTemplate.
+type SnippetsTemplateStatus struct {
+	// Conditions describes the state of the SnippetsTemplate.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// SnippetsTemplateConditionType is a type of condition associated with SnippetsTemplate.
+type SnippetsTemplateConditionType string
+
+// SnippetsTemplateConditionReason is a reason for a SnippetsTemplate condition type.
+type SnippetsTemplateConditionReason string
+
+const (
+	// SnippetsTemplateConditionTypeAccepted indicates that the SnippetsTemplate is accepted.
+	//
+	// Possible reasons for this condition to be True:
+	//
+	// * Accepted
+	//
+	// Possible reasons for this condition to be False:
+	//
+	// * Invalid
+	// * CRDInvalid
+	// * CRDNotFound
+	// * TargetNotFound
+	SnippetsTemplateConditionTypeAccepted SnippetsTemplateConditionType = "Accepted"
+
+	// SnippetsTemplateConditionReasonAccepted is used with the Accepted condition type when
+	// the condition is true.
+	SnippetsTemplateConditionReasonAccepted SnippetsTemplateConditionReason = "Accepted"
+
+	// SnippetsTemplateConditionReasonInvalid is used with the Accepted condition type when
+	// SnippetsTemplate is invalid.
+	SnippetsTemplateConditionReasonInvalid SnippetsTemplateConditionReason = "Invalid"
+
+	// SnippetsTemplateConditionReasonCRDInvalid is used with the Accepted condition type when
+	// the referenced CRD is invalid.
+	SnippetsTemplateConditionReasonCRDInvalid SnippetsTemplateConditionReason = "CRDInvalid"
+
+	// SnippetsTemplateConditionReasonCRDNotFound is used with the Accepted condition type when
+	// the referenced CRD is not found.
+	SnippetsTemplateConditionReasonCRDNotFound SnippetsTemplateConditionReason = "CRDNotFound"
+
+	// SnippetsTemplateConditionReasonTargetNotFound is used with the Accepted condition type when
+	// the target is not found.
+	SnippetsTemplateConditionReasonTargetNotFound SnippetsTemplateConditionReason = "TargetNotFound"
+)
+```
+
+#### Templates
+
+A template must be a [go template](https://pkg.go.dev/text/template).
+
+NGF gives a template access to the data and metadata about a Custom resource (corresponds
+to `SnippetsTemplate.spec.valuesCRD`) via `TemplateContext` type:
+
+```go
+// TemplateContext provides context for NGINX configuration templates.
+type TemplateContext struct {
```

Comment:
I'm not sure how this type fits in. Will NGF be unmarshaling the CRD referenced by `valuesCRD` into this type?

Reply:
>  Will NGF be unmarshaling the CRD referenced by valuesCRD into this type?

yep

Reply:
Can you add that detail to this section?